experiments/process_sandbox/library_runner.cc

```diff
@@ -239,184 +219,531 @@ namespace sandbox
+    size_t large_class = std::max(size_bits, SUPERSLAB_BITS) - SUPERSLAB_BITS;
+    return reserve_committed(large_class);
+  }
+  void* MemoryProviderProxy::reserve_committed(size_t large_class) noexcept
+  {
+    return reinterpret_cast<void*>(requestHostService(
+      MemoryProviderReserve, static_cast<uintptr_t>(large_class)));
+  }
+
+}
+
+namespace
+{
+  /**
+   * The function from the loaded library that provides the vtable dispatch
+   * for functions that we invoke.
+   */
+  void (*sandbox_invoke)(int, void*);
+
+  /**
+   * The run loop.  Takes the public interface of this library (effectively,
+   * the library's vtable) as an argument.  Exits when the upcall depth changes
+   * after executing the helper function.  This provides a nested runloop
+   * abstraction similar to OpenStep's modal runloop. Each recursion depth in
+   * an upcall has its own runloop that handles recursive invocations from the
+   * parent in response to the upcall.
+   */
+  void runloop(int upcall_depth = 0)
+  {
+    int new_depth;
+    do
+    {
+      do
+      {
+        if (shared->should_exit)
+        {
+          exit(0);
+        }
+      } while (!shared->token.child.wait(INT_MAX));
+      SANDBOX_DEBUG_INVARIANT(
+        shared->token.is_child_executing,
+        "Child is executing when the parent thinks is is not");
+      int idx = shared->function_index;
+      void* buf = shared->msg_buffer;
+      shared->msg_buffer = nullptr;
+      try
+      {
+        if ((buf != nullptr) && (sandbox_invoke != nullptr))
+          sandbox_invoke(idx, buf);
+      }
+      catch (...)
+      {
+        // FIXME: Report error in some useful way.
+        printf("Exception!\n");
+      }
+      new_depth = shared->token.upcall_depth;
+      // Wake up the parent if it's expecting a wakeup for this upcall depth.
+      // The `upcall` function has a wake but not a wait because it is using
+      // the `wait` in this function, we need to ensure that we don't unbalance
+      // the wakes and waits.
+      if (new_depth == upcall_depth)
+      {
+        shared->token.is_child_executing = false;
+        shared->token.parent.wake();
+      }
+    } while (new_depth == upcall_depth);
+  }
+
+  SNMALLOC_SLOW_PATH
+  void bootstrap()
+  {
+    void* addr = nullptr;
+    size_t length = 0;
+    // Find the correct environment variables.  Note that libc is not fully
+    // initialised when this is called and so we have to be very careful about
+    // the libc function that we call.  We use the `environ` variable directly,
+    // rather than `getenv`, which may allocate memory.
+    //
+    // The parent process provides the shared memory object in the file
+    // descriptor with the number given by `SharedMemRegion` and the location
+    // where it should be mapped in an environment variable.  The child has to
+    // map this as the first step in bootstrapping (before most of libc
+    // initialises itself) to get a working heap.
```

Comment:
It would be good to have an example of the syntax this accepts for the address and size.